### ISSUE TITLE
FJS support?

### ISSUE THREAD

**Issue by chroma04** [euwbah/musescore-xen-tuner#4]
[Functional Just Notation](https://misotanni.github.io/fjs/en/) is a relatively popular JI notation that uses fingering-like fractions to indicate a comma shift from a plain Pythagorean interval, instead of extra accidental symbols.

There are an infinite amount of unique commas (one for each prime) that can stack upon themselves and each other, so to fit them into one number they are all multiplied together: a fingering of **35/121** or (5\*7)/(11\*11) indicates a shift up by a 5- and 7-comma (80/81 and 63/64) and then a shift down by two 11-commas (33/32).

Since the numbers have to be parsed and factored into primes and their commas generated according to [their algorithm](https://misotanni.github.io/fjs/en/rules.html), it might be outside the scope of this project that mainly assists with managing accidentals... :stuck_out_tongue_winking_eye:

**Comment by euwbah** [comment]
Thanks for sharing about FJS. Hmm, I would like to implement this but you're right about this being a bit out of scope, as this requires a change to the internals. Right now, this plugin can parse JI ratios, but will tune them verbatim. However, this calls for JI ratios to represent commas w.r.t. the 3-limit chain of fifths, so a different operating mode of this plugin needs to be made. I'll mark this as a feature for later.

**Comment by euwbah** [comment]
PS i realise that the 'prime factorizing' ability can be artificially implemented using ligatures and text-based accidentals, since it could do a brute force search-and-replace for composite numbers and break them down into primes. So a python script could brute force all the powers of primes up to a certain required threshold.

Of course, for really large tunings, I don't think this will be feasible due to performance constraints, but you can check out how [HEJI](https://github.com/euwbah/musescore-xen-tuner/blob/master/tunings/heji/5%20limit.txt) is implemented.

E.g. (the comma up/down direction may be wrong/inverted)
```
A4: 440
b (2187/2048) #
'1/25' '1/5' (125/128) '5' '25'
'1/7' (63/64) '7'

lig(2,3)
-2 -1 '1/175'
-2 1 '7/25'
-1 -1 '1/35'
-1 1 '7/5'
//etc...
```

**Comment by Aumuse** [comment]
Hi! Inspired by this thread I have written a 255-odd-limit FJS tuning config (.txt), listing as secondary text accidentals (only, so the only primary accidental chain defined is Pythagorean) all the utonal (/n) then otonal (n) commas corresponding to each non-3 odd limit up to 253 (since 255 = 3×5×17 and can be represented with the 85 comma) and using FloraC's radius of tolerance, sqrt(2178/2048), instead of 65/63 - so the otonal 31-comma is 31/32 instead of 248/243 etc. (Credits to the calculator I used to obtain the commas: https://www.yacavone.net/fjs-explorer/?rotPreset=2)
Feel free to modify this config by adding more commas for higher odd limits, redefining the naturals and accidentals and so on :)
[255fjs.txt](https://github.com/user-attachments/files/16107123/255fjs.txt)

**Comment by Aumuse** [comment]
> Hi! Inspired by this thread I have written a 255-odd-limit FJS tuning config (.txt), listing as secondary text accidentals (only, so the only primary accidental chain defined is Pythagorean) all the utonal (/n) then otonal (n) commas corresponding to each non-3 odd limit up to 253 (since 255 = 3×5×17 and can be represented with the 85 comma) and using FloraC's radius of tolerance, sqrt(2178/2048), instead of 65/63 - so the otonal 31-comma is 31/32 instead of 248/243 etc. (Credits to the calculator I used to obtain the commas: https://www.yacavone.net/fjs-explorer/?rotPreset=2) Feel free to modify this config by adding more commas for higher odd limits, redefining the naturals and accidentals and so on :) [255fjs.txt](https://github.com/user-attachments/files/16107123/255fjs.txt)

With reference to this bug: https://github.com/euwbah/musescore-xen-tuner/issues/12#issue-2343512430, I have defined the fingering '1' as undoing all previous accidentals. Hence for instance F1 following F#23 is equal to F-natural-one instead of F-sharp-one, and if F#1 is meant to follow F#23, the sharp accidental must be repeated. This differs from the official recommendation that 1 only undoes a previous comma ratio, not the Pythagorean accidental as well.

**Comment by euwbah** [comment]
Oh wow @Aumuse, that's amazing. sorry it took me so long to see this. I've just fixed #12 so this should be no problem.

I should just write a python script to autogenerate these up to the 11-limit for primary accidentals + ligatures and N odd-limit using a custom radius of tolerance.

Just to clarify, sqrt(apotome) being the radius of tolerance means that the original chain-of-fifths note is within +/- sqrt(apotome) of the comma shifted one?

**Comment by euwbah** [comment]
Also, now that #13 is implemented (we can declare pythagorean accidentals to act independently of other accidentals), it appears that https://misotanni.github.io/fjs/en/crash.html#propagation-of-fjs-accidentals is supported, with `1` being the naturalizing accidental for all comma accidentals, while still preserving prior pythagorean accidentals.

Edit: I've just read the second example here: https://misotanni.github.io/fjs/en/rules.html#:~:text=The%20following%20four%20bars%20read%3A%20C%E2%99%AF5%20C%E2%99%AF5%3B%20C%E2%99%AF5%20C%E2%99%AF%3B%20C%E2%99%AF5%20C%3B%20C%E2%99%AF5%20C5

It appears that in FJS, the natural symbol/pythagorean symbols can override prior FJS numerical accidentals, but not the other way around. Different levels of propagation precedence is not yet supported, so if I let `1` be the naturalizing accidental for the FJS symbols only, then the natural symbol will allow prior FJS commas to fall through, and to fully naturalize a note one would need to spell it with a natural and `1`

**Comment by Aumuse** [comment]
> Oh wow [@Aumuse](https://github.com/Aumuse), that's amazing. sorry it took me so long to see this. I've just fixed [#12](https://github.com/euwbah/musescore-xen-tuner/issues/12) so this should be no problem.
> 
> I should just write a python script to autogenerate these up to the 11-limit for primary accidentals + ligatures and N odd-limit using a custom radius of tolerance.
> 
> Just to clarify, sqrt(apotome) being the radius of tolerance means that the original chain-of-fifths note is within +/- sqrt(apotome) of the comma shifted one?

Yes for the last question, if I understand the system correctly.

**Comment by euwbah** [comment]
Added fjs/255odd in commit 136b1eb0e20b557f36189a0add9b76f358683772
